core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -515,29 +528,32 @@ private[log] object LogCleanerManager extends Logging {
     * @param log the log
     * @param lastCleanOffset the last checkpointed offset
     * @param now the current time in milliseconds of the cleaning operation
-    * @return the lower (inclusive) and upper (exclusive) offsets
+    * @return OffsetsToClean containing offsets for cleanable portion of log and whether the log checkpoint needs updating
     */
-  def cleanableOffsets(log: Log, lastCleanOffset: Option[Long], now: Long): (Long, Long) = {
+  def cleanableOffsets(log: Log, lastCleanOffset: Option[Long], now: Long): OffsetsToClean = {
     // If the log segments are abnormally truncated and hence the checkpointed offset is no longer valid;
     // reset to the log starting offset and log the error
-    val firstDirtyOffset = {
+    val (firstDirtyOffset, needsUpdateCheckpoint) = {
       val logStartOffset = log.logStartOffset
       val checkpointDirtyOffset = lastCleanOffset.getOrElse(logStartOffset)
 
       if (checkpointDirtyOffset < logStartOffset) {
         // Don't bother with the warning if compact and delete are enabled.
-        if (!isCompactAndDelete(log))
+        if (!isCompactAndDelete(log)) {
           warn(s"Resetting first dirty offset of ${log.name} to log start offset $logStartOffset " +
             s"since the checkpointed offset $checkpointDirtyOffset is invalid.")
-        logStartOffset
+          (logStartOffset, true)
```

Comment:
I think it should be ok to also return true (update checkpoint file) for "compact, delete" topics as well, even though we are not printing a warning for them. @junrao What do you think? It seems consistent to update checkpoint for all the cases here that set firstDirtyOffset to log start; unless there is a reason to only update checkpoint when we print the warning about resetting first dirty offset?

Reply:
The reason that we don't log a warn here is that checkpointDirtyOffset is expected to be less than logStartOffset for compact and delete topic. However, it does seem we should update dirtyOffset in this case.